Reorganize the Admin Menu

MikeAlhayek opened the pull request:
Fix https://github.com/OrchardCMS/OrchardCore/issues/11942
Fix #12710


## Legacy Mode
If you are not ready yet to update the menu in your project, you can enable the legacy format by adding this code to your startup project

```
AppContext.SetSwitch("LegacyAdminMenuNavigation", true);
```


## Consolidate All Settings Under 'Settings' Menu and Replace 'Configuration' with 'Tools'
![image](https://github.com/user-attachments/assets/0daa5968-7c30-428d-93aa-0ba48f5de3af)

## Moved the "Content definitions" menu to "Design" tab

![image](https://github.com/user-attachments/assets/47171708-608f-4e41-bcac-d51cdc94dd47)


## Access Control 
![image](https://github.com/user-attachments/assets/d1e0fdbc-a362-44d2-bad3-80cadf4de45c)

![image](https://github.com/user-attachments/assets/49f327b0-b07b-4427-852e-00bcad44f764)


## Settings Menu (All Features Enabled)
![image](https://github.com/user-attachments/assets/aa4794fb-8ea3-4d1b-b761-dfcf7ef7903d)

![image](https://github.com/user-attachments/assets/082cc7a0-7d27-413c-9b90-00b56b0adaa8)

![image](https://github.com/user-attachments/assets/42bb6c8c-8ad4-4fef-b49b-ff8fcf3bb158)

![image](https://github.com/user-attachments/assets/6bbc830f-f66c-4c1a-9435-490511e16856)

![image](https://github.com/user-attachments/assets/3c8526a0-f56a-43df-87c7-5061ce0b278b)

![image](https://github.com/user-attachments/assets/e5b36a94-c0bb-47f4-9a98-877d60a4e1eb)


## Tools Menu
I also moved the Workflows and Audit Trail under Tools instead of the root level.

![image](https://github.com/user-attachments/assets/cebd492e-3aba-4f86-8dc2-5b54d481883e)

![image](https://github.com/user-attachments/assets/1bade523-8abc-4b99-baa3-84e78b84c8c1)

![image](https://github.com/user-attachments/assets/5bea59e5-c340-4454-aba7-0a71004fa6f1)

![image](https://github.com/user-attachments/assets/15fffaa5-94e5-48d1-a7d0-957ca2334838)

![image](https://github.com/user-attachments/assets/4690cf13-8419-499d-80df-bf6d3583d914)

hishamco commented:
This is the first time I have seen `Tools` menu, where did it come from?!!

MikeAlhayek commented:
Please read the referenced issues

hishamco commented:
> Move any site setting or configurations under the “Configuration > Settings” menu

No `Tools` mentioned there?!! BTW the link in the related issue is broken

MikeAlhayek commented:
Look at the discussions. Tools is a better term to use here since we added the Settings. Do
You have an issue with Tools?

gvkries commented:
I like it, but I'm not sure about moving audit trail. I think it's important enough to stay at the root. 

While you are at it, do we agree to move GraphiQL to Tools, or should it go to search as requested in #12710? I think it makes sense in the new tools menu.

gvkries commented:
Additionally, can we somehow group authentication and security settings? They seem to be closely related IMHO.

MikeAlhayek commented:
> I like it, but I'm not sure about moving audit trail. I think it's important enough to stay at the root.

This was requested by @sebastienros in one of the meetings. I prefer keeping the main menu free for users to utilize as they see fit. While the importance of features is clear, it can vary greatly from one user to another. So, why should the Audit Trail take priority over other features like Workflows, the Admin Menu, Features or Deployments?

> Additionally, can we somehow group authentication and security settings? They seem to be closely related IMHO.

I do not disagree with you there and Authentication is part of security. I am just trying to not add yet another level to the depth of the menu. and is one of the reasons why I went down an [alternative approach here ](https://github.com/OrchardCMS/OrchardCore/issues/11942#issuecomment-2430196848) so that the main menu is much thinner and not too deep.

MikeAlhayek commented:
@gvkries I moved the `OpenId Connect` and `Authentication` under `Security` just so we can see if the experience is acceptable or not. Feel free to checkout the branch and test driving it.

I also renamed the "Security" tab at the root level to "Access Control." This feels like a much better fit, especially since the security settings have been moved to the Settings tab. The term "Security" can often feel overused, and "Access Control" provides a clearer, more specific representation of the functionality related to users, roles, and authentication.

Piedone commented:
In general, I'm OK with this.

However, I see frustration both for users (to learn and unlearn the new menu structure) and developers (basically all admin menus will need to be checked, most changed) when getting this update. Documentation needs to be updated all over the place as well. All of these are subtle breaking changes one will only discover when bumping into an issue or users complaining, not instantly clear like a build error.

I don't know how to make this easier. For sure it starts with documenting it, and providing a search & replace pattern for the simple case of Configuration -> Settings, so you can at least quickly fix that in the codebase.

MikeAlhayek commented:
Once we all agree on the new menu, I'll update the documentation and add release notes updates. We'll probably talk more about it during triage tomorrow. 

I do agree it's going to require every project to change. But the good news that it's part of 3.0 :)

sebastienros commented:
If we could not screw users for some time that would be great. There are breaking changes that don't break anyone or are super natural to fix (`[Obsolete]` with what to do instead) and breaking changes that really suck because they seem like unnecessary small changes and impact lots of installs.

> But the good news that it's part of 3.0 :)

This doesn't sound good. It's not because we had to ship some breaking changes that we can add any other ones. Every breaking change should be done with measure.

gvkries commented:
I've tried this out and I love it :-) 

Some things I would like to mention:

- Features: I've expected this in Settings first, but it kind of makes more sense in tools. Might need to get used to it, because it was configuration before.
- As noted above in a comment, I would stick with the existing title casing. I find that more compelling.
- Keeping tools and settings at the end makes sense for me. E.g. moving multi-tenancy up.

Thanks @MikeAlhayek

MikeAlhayek commented:
> I've tried this out and I love it :-)

This is awesome!

>I would stick with the existing title casing. I find that more compelling.
Personally I think we are abusing the Title case in the admin side. It feels like everything is a Title-Case "but at the same time we are not consistent currently not even in the admin menu". For me it's important to stay consistent. Title case should be more for headlines but not for menu or buttons. Title case should be used to highlight something important. I think sentence-case is much more appropriate for the navigation. I think we should see what others also think and then I'll make that bulk change if needed at the end. I also did a quick Google search and got this

![image](https://github.com/user-attachments/assets/0be3b992-3834-4052-a8f3-f42fce74de22)


> Keeping tools and settings at the end makes sense for me. E.g. moving multi-tenancy up.

Done

MikeAlhayek commented:
Would love to hear your thoughts on the new structure of the admin menu! Please share your feedback with us.
@agriffard @BenedekFarkas @deanmarcussen @davidhayden @alexbocharov @barthamark @domonkosgabor @kevinchalet @ns8482e @sarahelsaig @sfmskywalker @Skrypt @SzymonSel @jptissot @urbanit @hyzx86 @MichaelPetrinolis @giannik @infofromca @TFleury @netwavebe @carlwoodhouse @brunoAltinet @MarGraz @PBMikeW @DrewBrasher @vengi83644 @AndreySurkov 

Others? Everyone is welcome :)

ns8482e commented:
Before this gets merged - we should have guideline on menus -

- What's the purpose of the top level menus?
- What makes that menu  top level?
- What kind of functionality should go within it?
- How a custom module should contribute to these menu?
- What are the things Module author should consider while contributing to the menus? So their modules stay relevant in these kind of changes in future


If we don't provide guidelines for future ourselves then, we might call these new menu structures legacy and may attempt to change again.

MikeAlhayek commented:
@ns8482e I agree. This is going to be part of the release notes. Once we finalize these changes, I'll be updating the docs and adding the release notes.

ns8482e commented:
I personally don't like such breaking changes! When there is not value addition - as menu reorg is fully supported out of the box without making change in OC's code base 

The value addition would be around this feature is to allow user to reorg menus from within Admin UI

ns8482e commented:
Also with this change - there is no value addition orchard core broader user base those are used to existing menu system and don't see any difficulty currently using it.  

For agencies that needs new menu system for their user base can build on new menu system using Navigation framework with their own menu name and replaced in ther Layout - the scenario is supported out of the box.

Also adding things related to UI in app-setting makes app setting more noisy.


If OC really needs new menu system for 3 and owards - I would suggest not to make any change to AdminMenu.cs, instead create separate files for like `ToolsAdminMenu.cs`,  `SettingsAdminMenu.cs` and  register them as provider with seperate menu name

ns8482e commented:
Value addition around restructuring menu would be, 

1. Add Menu Rule provider - like `menu.json`
2. Evaluate the rule for given menu, and find position from provider json
3. Place the menu according to the rule.

ns8482e commented:
> @ns8482e I agree. This is going to be part of the release notes. Once we finalize these changes, I'll be updating the docs and adding the release notes.

I meant not the release note - my comment was to have guideline (pre-condition of this PR) built and publish in docs which the PR should follow

MikeAlhayek commented:
@gvkries one thing that still feels off to me is the fact that "Content Definition" menu is located under "Content". This menu should be moved either under Tools or as suggested by @sebastienros under Design menu. Design may not be a good home for it as these are design tool for content structures. Thoughts on re-homing it?

MikeAlhayek commented:
@Piedone @sebastienros @gvkries I think another good option is to move "Media" to a root level away from "Content". Under Media we can have Library and Profiles. I don't media profiles belong under settings menu.

gvkries commented:
> @gvkries one thing that still feels off to me is the fact that "Content Definition" menu is located under "Content". This menu should be moved either under Tools or as suggested by @sebastienros under Design menu. Design may not be a good home for it as these are design tool for content structures. Thoughts on re-homing it?

In general I think "content definition" makes sense in "content" as it is. But for me this one can also be located in "design", i.e. "designing a content type". What I don't like is e.g. "workflows" in design, because for me this has nothing directly to do with designing the site.
I think we should be careful not putting everything in "Tools", e.g. I agree that media profiles should be in media. But I think it can stay in "content" even with the profiles.

gvkries commented:
Thinking more about media profiles, I would keep this in "settings" with the other media options. It's a one time configuration for a site, not really touched often after everything is "set up".

gvkries commented:
And by following my reasoning for media profiles, I think "content definition" may benefit from moving it out of "Content".

MikeAlhayek commented:
@gvkries I'll move Content Definition to Design as you, @sebastienros and myself seem to agree.

I still strongly believe that Media is not really content and should be reallocated. I think having Profile and Library in a root menu is better that placing it in the content. Profiles is something we can create even after the site is created. It's not really a settings. I want to move it out to get a feel of how it will feel and hopefully we get more feedback from others as well.

gvkries commented:
@MikeAlhayek I'm fine with that 🤗

vengi83644 commented:
I am happy with these changes. The current menu is hard to understand at first glance, even though we can search through it.

MikeAlhayek commented:
I love having the Media on the root level

![image](https://github.com/user-attachments/assets/9ca71925-6927-486a-8e98-4a2abbc749df)

Also moving Content definition to Design is much better for me.
![image](https://github.com/user-attachments/assets/64a1c105-6472-4ad1-80f8-29b0eb1c9351)

MikeAlhayek commented:
@gvkries if you can, please checkout the PR and test drive the changes.

Here is a recipe that would enable all the feature that I am using in you or someone else want to test drive

```JSON
{
  "steps": [
    {
      "name": "Feature",
      "enable": [
        "Application.Default",
        "OrchardCore.Cms.Web",
        "OrchardCore.Settings",
        "OrchardCore.Admin",
        "OrchardCore.Liquid",
        "OrchardCore.Contents",
        "OrchardCore.ContentTypes",
        "OrchardCore.Shortcodes",
        "OrchardCore.Html",
        "OrchardCore.Recipes.Core",
        "OrchardCore.Title",
        "OrchardCore.AdminMenu",
        "OrchardCore.Templates",
        "OrchardCore.AdminTemplates",
        "OrchardCore.Alias",
        "OrchardCore.Apis.GraphQL",
        "OrchardCore.ArchiveLater",
        "OrchardCore.AuditTrail",
        "OrchardCore.Autoroute",
        "OrchardCore.Scripting",
        "OrchardCore.Recipes",
        "OrchardCore.BackgroundTasks",
        "OrchardCore.ContentFields",
        "OrchardCore.ContentFields.Indexing.SQL",
        "OrchardCore.Roles.Core",
        "OrchardCore.Users",
        "OrchardCore.ContentFields.Indexing.SQL.UserPicker",
        "OrchardCore.Localization",
        "OrchardCore.Users.Localization",
        "OrchardCore.ContentLocalization",
        "OrchardCore.ContentLocalization.ContentCulturePicker",
        "OrchardCore.Sitemaps",
        "OrchardCore.ContentLocalization.Sitemaps",
        "OrchardCore.ContentPreview",
        "OrchardCore.Deployment",
        "OrchardCore.Contents.Deployment.Download",
        "OrchardCore.Contents.Deployment.ExportContentToDeploymentTarget",
        "OrchardCore.CustomSettings",
        "OrchardCore.Deployment.Remote",
        "OrchardCore.Diagnostics",
        "OrchardCore.DynamicCache",
        "OrchardCore.Email",
        "OrchardCore.Email.Azure",
        "OrchardCore.Email.Smtp",
        "OrchardCore.Facebook",
        "OrchardCore.Users.ExternalAuthentication",
        "OrchardCore.Facebook.Login",
        "OrchardCore.Facebook.Pixel",
        "OrchardCore.Widgets",
        "OrchardCore.Facebook.Widgets",
        "OrchardCore.Resources",
        "OrchardCore.Features",
        "OrchardCore.Feeds",
        "OrchardCore.Flows",
        "OrchardCore.Forms",
        "OrchardCore.GitHub.Authentication",
        "OrchardCore.Google.Analytics",
        "OrchardCore.Google.GoogleAuthentication",
        "OrchardCore.Google.TagManager",
        "OrchardCore.HealthChecks",
        "OrchardCore.HomeRoute",
        "OrchardCore.Https",
        "OrchardCore.Indexing",
        "OrchardCore.Rules",
        "OrchardCore.Layers",
        "OrchardCore.Lists",
        "OrchardCore.Localization.AdminCulturePicker",
        "OrchardCore.Localization.ContentLanguageHeader",
        "OrchardCore.Markdown",
        "OrchardCore.Media",
        "OrchardCore.Media.Cache",
        "OrchardCore.Media.AmazonS3",
        "OrchardCore.Media.AmazonS3.ImageSharpImageCache",
        "OrchardCore.Media.Azure.ImageSharpImageCache",
        "OrchardCore.Media.Azure.Storage",
        "OrchardCore.Media.Indexing",
        "OrchardCore.Media.Indexing.OpenXML",
        "OrchardCore.Media.Indexing.Pdf",
        "OrchardCore.Media.Indexing.Text",
        "OrchardCore.Media.Security",
        "OrchardCore.Media.Slugify",
        "OrchardCore.Menu",
        "OrchardCore.Microsoft.Authentication.AzureAD",
        "OrchardCore.Microsoft.Authentication.MicrosoftAccount",
        "OrchardCore.Navigation",
        "OrchardCore.Notifications",
        "OrchardCore.Notifications.Email",
        "OrchardCore.Sms",
        "OrchardCore.Notifications.Sms",
        "OrchardCore.OpenId",
        "OrchardCore.OpenId.Client",
        "OrchardCore.OpenId.Management",
        "OrchardCore.OpenId.Server",
        "OrchardCore.OpenId.Validation",
        "OrchardCore.Placements",
        "OrchardCore.PublishLater",
        "OrchardCore.Queries.Core",
        "OrchardCore.Queries",
        "OrchardCore.Queries.Sql",
        "OrchardCore.ReCaptcha",
        "OrchardCore.ReCaptcha.Users",
        "OrchardCore.XmlRpc",
        "OrchardCore.RemotePublishing",
        "OrchardCore.ReverseProxy",
        "OrchardCore.Roles",
        "OrchardCore.Search",
        "OrchardCore.Search.Elasticsearch",
        "OrchardCore.Search.Elasticsearch.ContentPicker",
        "OrchardCore.Search.Elasticsearch.Worker",
        "OrchardCore.Security",
        "OrchardCore.Seo",
        "OrchardCore.Shortcodes.Templates",
        "OrchardCore.Sitemaps.Cleanup",
        "OrchardCore.Sitemaps.RazorPages",
        "OrchardCore.Sms.Azure",
        "OrchardCore.Taxonomies",
        "OrchardCore.Taxonomies.ContentsAdminList",
        "OrchardCore.Themes",
        "OrchardCore.Twitter",
        "OrchardCore.Twitter.Signin",
        "OrchardCore.UrlRewriting",
        "OrchardCore.Users.2FA",
        "OrchardCore.Users.2FA.AuthenticatorApp",
        "OrchardCore.Users.2FA.Email",
        "OrchardCore.Users.2FA.Sms",
        "OrchardCore.Users.AuditTrail",
        "OrchardCore.Users.ChangeEmail",
        "OrchardCore.Users.CustomUserSettings",
        "OrchardCore.Users.Registration",
        "OrchardCore.Users.ResetPassword",
        "OrchardCore.Users.TimeZone",
        "OrchardCore.Workflows",
        "OrchardCore.Workflows.Http",
        "OrchardCore.Workflows.Session",
        "OrchardCore.Workflows.Timers",
        "SafeMode",
        "TheAdmin",
        "TheBlogTheme"
      ]
    }
  ]
}
```

Piedone commented:
"Design" is very much about look and feel for me, so I wouldn't put Content Definition there; that's good under Content (I wouldn't move it, but if everybody is against it staying, then at least Settings). Media, however, I'd consider either content (it is _content_, though it's not _content items_) or a top-level item.

However, it has always bothered me that we have Content -> Content Types, and it's not actually about content types (and not to be confused with Content -> Content Definition -> Content Types), but content items filtered by content types. Perhaps "Items by Type"?

sebastienros commented:
@Piedone hopefully you can watch the meeting soon

https://dictionary.cambridge.org/dictionary/english/design

MikeAlhayek commented:
@Piedone 
> "Design" is very much about look and feel for me

That is what I thought at first. However, it contains other things like shortcodes and placements. If these were not there, they I would have suggested it to name it "Appearance" instead of "Design". That said, "Design" is more than Appearance, so we thought its better to move the content definition "which is a content design" and keep the menu called "Design".

@sebastienros even suggested to move Workflows into Design too which I don't mind. But @gvkries did not like that idea. Thoughts on that? Workflow can fit under tools and under Design.

>  it has always bothered me that we have Content -> Content Types, and it's not actually about content types

This is part of the blog recipe and can be changed. It's not part of the default navigation providers. We can always put that one level up instead of it being in "Content types" as well.

> Media, however, I'd consider either content (it is content, though it's not content items) or a top-level item

It is currently on the top level and I find it much better this way. So we agree on the media being top level.

Piedone commented:
OK then! Please fix the Blog recipe too, then.

Workflows is much more Tools for me than Design.

MikeAlhayek commented:
@Piedone I updated the blog recipe. I am not sure why we were showing all content types on the menu. Not very helpful. It was just extra clutter. I updated it to only show Articles, Pages, and Taxonomy.

![image](https://github.com/user-attachments/assets/ba0c797a-9ab8-419a-aaa8-05bfbf7c20a9)

Piedone commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/AdminMenu.cs:
```diff
@@ -40,12 +40,12 @@ protected override ValueTask BuildAsync(NavigationBuilder builder)
             .Add(S["Design"], content => content
                 .Add(S["Content definition"], S["Content definition"].PrefixPosition(), contentDefinition => contentDefinition
-                    .Add(S["Content types"], S["Content types"].PrefixPosition("1"), contentTypes => contentTypes
+                    .Add(S["Types"], S["Types"].PrefixPosition("1"), types => types
```
I'd keep this and "Content parts". Yes, it's more verbose, but that doesn't matter and it's also clearer, because "type" and "part" are overloaded terms.

MikeAlhayek replied:
I actually removed all the duplicates terms.  In this case, the parent menu is Content definition. So type and parts is much simpler and cleaner than having to repeat context again "Content Types" and "Content Parts".

Piedone replied:
I don't think cleanliness or simplicity is a priority over clarity here.

MikeAlhayek replied:
I think it's clear. But no strong opinion. If you have strong opinion on being repeating context, I'll prepend the word "Content".

Piedone replied:
My opinion is strong on keeping these links :). It's not really repeating: you and I and everybody in this PR understand this, but what is listed under e.g. Content definition -> Types, are "content types", not "content definition types", or simply "types".

OC users have to understand and follow a bunch of terminology, let's make this easier whenever possible. (And as a bonus, not have yet another new translation string.)

MikeAlhayek replied:
done

Piedone replied:
Thanks!

Piedone commented:
Looks good!

MikeAlhayek commented:
@Piedone @sebastienros @gvkries Any objections to merging this PR? I'd like to give everyone some time to familiarize themselves with the new structure and potentially make any necessary updates or suggest further improvements we might have missed. Once we release a new preview with these changes, I hope it will prompt more feedback (if any) from the community.

Piedone commented on src/OrchardCore.Modules/OrchardCore.GitHub/AdminMenuGitHubLogin.cs:
```diff
@@ -34,6 +36,24 @@ protected override ValueTask BuildAsync(NavigationBuilder builder)
+        builder
+            .Add(S["Settings"], settings => settings
+                .Add(S["Security"], S["Security"].PrefixPosition(), security => security
+                    .Add(S["Authentication"], S["Authentication"].PrefixPosition(), authentication => authentication
```
Rather External Authentication, for the others too.

MikeAlhayek replied:
I would rather keep it Authentication. Why would we call it external authentication where is this the other authentication item we have in the menu. The other thing have is Access control

Piedone replied:
The only items under this after enabling every feature are external authentication settings.

MikeAlhayek replied:
Authentication would shorter with same meaning. Also if we ever need to add authentication settings related to local authentication that would go under an authentication otherwise we could end up with Authentication and External Authentication.

Piedone replied:
Authentication is also about local authentication, the settings of which (related to login and registration) are not under this menu but a level up.

And yes, we could totally have a differentiation here, with a Local Authentication and External Authentication menu (or just one Authentication, but then having all authentication-related things under it).

Additionally, now we have Access Control, Security, and Authentication, which is confusing.

MikeAlhayek replied:
**Access Control** is for managing access. Security and Authentication items under the settings menu are are different things.
Security are security related settings for the app itself like header, CORS, https..

If you think we should call it External Authentication, then maybe we should also move the local authentication under "Authentication" so we have two sub menus.

MikeAlhayek replied:
Looking more at the current setup, I think Registration, User Login, and Change Email are great under "Security". External Authentication providers are under "Authentication" make sense to me without the word "External". If you want to suggest a new setup, please share a bullet bases structure so the picture becomes clear. The following looks great to me.

![image](https://github.com/user-attachments/assets/d2ce179b-a1fe-4388-bdec-29ed8c29618a)

gvkries replied:
I also think that we don't need to have an extra "external" authentication. External vs. local is still all "Authentication" and easy to find there.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Email/AdminMenu.cs:
```diff
@@ -24,22 +24,51 @@ public AdminMenu(IStringLocalizer<AdminMenu> stringLocalizer)
+                          .Permission(Permissions.ManageEmailSettings)
+                          .LocalNav()
+                        )
+                       .Add(S["Email test"], S["Email test"].PrefixPosition(), entry => entry
```
It's a tool to test email or SMS service.

Piedone replied:
Arguably yes, but nobody will think about it like this when setting up e-mail (or SMS) delivery. You want to test e-mail sending after changing the e-mail config.

Piedone replied:
Your first thought would be to look four steps elsewhere in the menu structure, under an item named "Testing"? Perhaps if you configure e.g. SMTP settings in appsettings (i.e. not under Settings -> Communication -> Email) then it's better, but it's not at all intuitive when working with the settings from the admin. A notification displayed after saving the e-mail/SMS settings along the lines of "Do you want to verify if these settings work? Send a test email <here>." 

"Testing" is not my favorite in itself, I'd first understand it as having something to do with automated testing like unit tests. Having a Tools -> Communication submenu too seems better.

MikeAlhayek replied:
Testing would be for any kinds of testing. At least that was the idea. So if we have more service test tools, we can place them under the testing.

Piedone replied:
If we'll ever have those, most possibly they'll more appropriate in a more expressive submenu too.

MikeAlhayek replied:
I feel "Testing" menu here is a great fit. If you have a batter generic name, I am okay with it. But these should not be moved into settings. 


At a high level, any UI that users OrchardCore.Settings should be put in Settings along with readonly settings like the media options. Any tool like testing, url rewriting, features, etc would be a tool.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/AdminMenu.cs:
```diff
@@ -18,15 +18,34 @@ public AdminMenu(IStringLocalizer<AdminMenu> stringLocalizer)
+
         builder
-            .Add(S["Content"], content => content
-                .Add(S["Content Definition"], S["Content Definition"].PrefixPosition("9"), contentDefinition => contentDefinition
```
The problem is that we are overusing title case in OC. Based on best practices, menus are usually sentence case not title case. Title case is usually for things that you want to stand out "not to be used everywhere".

Fortunately in OC, most of the items in our menu are one word and won't have much changes.

MikeAlhayek replied:
I made this change to use Title Case. There was few items that were sentence case, I changed them to title case at least so we are consistent.

Piedone replied:
Thanks!

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Features/AdminMenu.cs:
```diff
@@ -22,8 +22,22 @@ public AdminMenu(IStringLocalizer<AdminMenu> stringLocalizer)
+
         builder
-            .Add(S["Configuration"], configuration => configuration
+            .Add(S["Tools"], tools => tools
```
Are you suggesting to move Features and Admin Menu to settings?

MikeAlhayek replied:
I strongly disagree with your suggestion to moving Feature and Admin Menu to the Settings menu. I think they should stay under tools. These are better fit for tools. + we don't need to treat everything are settings and clutter the settings menu more.

Piedone replied:
Why is toggling feature states more of a tool than configuration?

MikeAlhayek replied:
It's a tool to enable/disable features. It's not purely settings.  We should not just start moving things into settings just because they make changes. We should only put settings related items under settings.

Piedone replied:
Isn't the configuration page to change the site's title also but a tool to set the title? Or would a different features UI, like with a list of checkboxes for each feature and a Save button at the bottom, make it more settings-y? :)

MikeAlhayek replied:
To me, if we have a form to fill out or change settings, that would be under the Settings. Features, its a tool to manage features.

Piedone replied:
What if I tell you that the Features page is actually a form already? :)

My point is, we have to distinguish between these based on the underlying principle and not a perception of the current UI, and toggling feature states is changing configuration, i.e. settings, in the same way as changing the site title is.

MikeAlhayek replied:
If you strongly feel it should be in Settings while I strongly feel it should be in Tools, I think we should get other opinions on this. We'll discuss it more in the Steering meeting today.

MikeAlhayek replied:
@Piedone talked about this in the meeting. @sebastienros  and @gvkries suggested to keep it in Tools as well.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminMenu.cs:
```diff
@@ -22,18 +22,34 @@ public AdminMenu(IStringLocalizer<AdminMenu> stringLocalizer)
     protected override ValueTask BuildAsync(NavigationBuilder builder)
     {
-        builder
+        if (NavigationHelper.UseLegacyFormat())
```
Planning to update the docs before merging. Just need to finalize the PR first.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Facebook/AdminMenu.cs:
```diff
@@ -21,10 +21,28 @@ public AdminMenu(IStringLocalizer<AdminMenu> stringLocalizer)
-                .Add(S["Settings"], settings => settings
-                    .Add(S["Meta App"], S["Meta App"].PrefixPosition(), metaApp => metaApp
+            .Add(S["Settings"], settings => settings
+                .Add(S["Integrations"], S["Integrations"].PrefixPosition(), integrations => integrations
```
"Social Meta Settings" is a custom settings that is part of the Seo module. All custom site settings are placed directly under the Settings menu.

Piedone replied:
I see.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Media.AmazonS3/AdminMenu.cs:
```diff
@@ -14,9 +14,24 @@ public AdminMenu(IStringLocalizer<AdminMenu> stringLocalizer)
+            return ValueTask.CompletedTask;
+        }
+
+        builder.Add(S["Tools"], tools => tools
```
Agreed. I moved them into Settings.

Piedone commented:
https://github.com/OrchardCMS/OrchardCore/pull/17112#discussion_r1887780873 brings up a larger issue: the majority of this structuring is opinionated and not at all clear to multiple people in the same way. I'm afraid we'll go circles here, and the end result won't make more people happy than where we started.

MikeAlhayek commented:
> https://github.com/OrchardCMS/OrchardCore/pull/17112#discussion_r1887780873 brings up a larger issue: the majority of this structuring is opinionated and not at all clear to multiple people in the same way. I'm afraid we'll go circles here, and the end result won't make more people happy than where we started.

Everything that we have today about the UI is opinionated. I am not sure there is a new issue.

Piedone commented:
The new issue is that the current menu is already out there, and has been for a long time, while this PR moves things around, forcing everyone, developers and users alike, to adapt, and in a significant way. Our conversations in the various comment threads and otherwise show that this is not at all a clear matter, and thus we'll change the current opinion with a different opinion, thinking that it'll be better, but ultimately not having much to back it. UX is not a hard science, but it can still be more methodical than developers sharing their opinions. And in such a case, the status quo wins, because it doesn't need everyone to put work into it.

I don't think we should change the menu after conversing just among ourselves. I know, and commend, that you've asked for wider feedback, but us not getting any doesn't mean that most people will be happy with the result. I see no point in arguing among ourselves further about things like whether something is a tool or a setting. Perhaps we're the worst people for this, actually, since it has been very long since we needed to find something in the menu without having a clue, or without understanding the fundamentals.

So, taking a step back, I propose rethinking the admin experience, something that's not a new idea by far: with the help of a UX professional (I have ideas on how to find them and Lombiq can fund their work), do the following:

1. Find a suitable open-source Boostrap dashboard theme/template as a base. We'll need to build a somewhat custom admin story, but not from the ground up.
2. Figure out the basic Orchard Core-specific UI principles the admin should follow.
3. Update a single UI to be our perfect role model (like the basic content item editor), and the menu structure. The rest of the UI will be kept more or less as-is, but with updated common visuals (coming from the theme) and quick wins (e.g. common pieces like buttons, hints, labels, fields, will all be updated just by updating the common classes, or with search and replace).
4. Release it as part of a major version. In the future, perhaps, maybe, follow the role model in all existing UI, and definitely in new UI. But whatever we'll have will be good enough.

BTW @BenedekFarkas re https://github.com/OrchardCMS/OrchardCore/issues/11942#issuecomment-2554214078: I thought about menu item favoriting too. It's perhaps something that an existing template can offer, but if not, it's not a big deal to implement.

MikeAlhayek commented:
@Piedone At this point, don't you think restarting the process might be too late, given the significant time we've already spent discussing it across various platforms?

If you have the time to consult a UI expert to guide the menu structure, I'm all for it. However, I also want to avoid letting this PR sit for too long, especially considering the amount of time and effort already invested by everyone, including yourself.

The current structure in this PR reflects the efforts to incorporate all the feedback we've received, aiming to create a solution that works for the majority.

MikeAlhayek commented:
> the changes are disruptive

Its not. We already provide a backward compatible (legacy mode).

> the changes are disruptive, so we better be sure of it, lest we have to repeat it. 

Is there a timeline on when you'll have a UX expert look at this and provide us feedback?

Piedone commented:
It's disruptive because while you can switch back temporarily, we surely won't maintain two menus forever. At that point, it'll still come at once, it's not something gradual, just optionally delayed.

I don't see much point in moving around the menu in itself, because we're missing the forest from the trees: What we need is a path to improve the admin experience, coherently and with guiding principles, not small pieces of it in isolation and without an overarching plan. Nobody will be able to help us with just the menu, without thinking through the whole thing: we might be better off with a different structure of the admin area, to begin with, perhaps with other ways of navigation as well, even secondary as you previously suggested, or whatever else. 

This will take months, but we'll have something for the long-term. If we want to try this approach, I can look into finding a suitable UX person now, and with the holidays I'd guess start this in earnest mid-January.

MikeAlhayek commented:
> It's disruptive because while you can switch back temporarily

It's temporary if we decide it going to be temporary. It could be a long term or permanent placement.


> I don't see much point in moving around the menu in itself, because we're missing the forest from the trees

While you don't, I and many other see the point.


> This will take months, but we'll have something for the long-term. If we want to try this approach, I can look into finding a suitable UX person now, and with the holidays I'd guess start this in earnest mid-January.

I feel that you are purposely trying to complicating this just because you don't seem to agree with the request.

The request is coming from many feedback I personally get. And others, tested it and seem to be happier with it. 

Anyway, will talk more about this during the next meeting, hopefully you can join.

Piedone commented:
I [also told in the very beginning](https://github.com/OrchardCMS/OrchardCore/issues/11942#issuecomment-2412110991) that:

> We should go outside of the circle of a few core maintainers and see what users say. Perhaps also ask some UX person and see what other systems do.

I'm still suggesting the same.

It's not that I merely disagree (I personally actually agree with almost all the changes here, as you've seen); rather, I'm of the viewpoint that these changes, in this form, aren't a net benefit to OC itself and the wider ecosystem.

MikeAlhayek commented:
> I'm still suggesting the same.

I understand. But, unless you are willing to find a consultant to provide us with their opinion, no one will. We already tried to ping the community to get as much feedback as possible. I internally consulted the end users "not developers" as well. So we have been following your suggestion all along and the current state of the PR seems to be accepted by folks other than me like @sebastienros @gvkries and other community members and internal end users.

Piedone commented:
I am willing, that's what I told above, and Lombiq will fund it.

What was the feedback those internal end users had to share about the current menu?

MikeAlhayek commented:
> I am willing, that's what I told above, and Lombiq will fund it.

Can't wait to hear what feedback we get back. Please update us on the process. Hopefully it won't be long time before your consultant some back with some feedback. 

> What was the feedback those internal end users had to share about the current menu?

They love the current state. They said the menu is much easier to navigate and find what they are looking for. I incorporated the feedback previously which was mainly around settings.

Piedone commented:
From the meeting today:

- I'll update the community about where the initial discussion with UX people go; I've sent out a couple of e-mails and submitted an ad to https://opensourcedesign.net/jobs/. As we talked about, the revamp of the admin area would be a long and painful process. I think it is necessary, though, and more importantly, to have an organized instead of an ad-hoc approach to building the admin UI.
- As for this PR, while I uphold the viewpoint elaborated above, I won't go against the vote of the majority.

MikeAlhayek commented:
@Piedone thank you.

I’d suggest opening a separate issue for the admin theme overhaul so we can dive into that more specifically and keep track of actionable items there. Anything related to the admin menu can stay in this PR so we don’t mix too many things up, especially since this PR is just focused on one part of the overall UX.

Piedone commented:
Follow-up issue: https://github.com/OrchardCMS/OrchardCore/issues/17327.

gvkries commented:
What about making this opt-in for now and maybe switch that later on (still to decide). This way we don't lose this great work and we can still change it, whenever there is some conclusion about the admin portal at all.

gvkries commented:
> @gvkries I honestly think this should be merged as-is then worry about completing the admin portal idea whenever that is concluded and not muddy the water with two different initiatives.

I agree with you, I just thought that we could accommodate others by making this optional for the time being :)

MikeAlhayek commented:
By making this the default option tells everyone this is the way forward. Anyone is just starting to use OC will get used to the new way. Anyone that wants to stick with the legacy mode, they can set it in their own project. But I think the mode should be supported for a very long time (or at least until the new portal is completed).

@Piedone stated if the majority want to merge it, go ahead.  I am not sure what that means but I think if we get couple more approvals, I think we should merge it as-is.

Piedone commented:
It means I'm neither approving nor disapproving, up to you.

pourdelir commented:
I am not a professional developer, althogh I have IT backgrounf. 
I am an enduser in this case.  I quiet liked Orchard, however, it is very confusing to set it up. the installation was straigt forward. I need more examples and trainings to get the job done.
